File: ios/Classes/controller/RtpSenderController.swift
import Flutter

/// Controller of an `RtpSender`.
class RtpSenderController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Instance of the `RtpSender`'s proxy.
  private var rtpSender: RtpSenderProxy

  /// ID of the channel created for this controller.
  private var channelId: Int = ChannelNameGenerator.nextId()

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Initializes a new `RtpSenderController` for the provided `RtpSenderProxy`.
  init(messenger: FlutterBinaryMessenger, rtpSender: RtpSenderProxy) {
    let channelName = ChannelNameGenerator.name(
      name: "RtpSender",
      id: self.channelId
    )
    self.messenger = messenger
    self.rtpSender = rtpSender
    self.channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: messenger
    )
    self.channel.setMethodCallHandler { call, result in
      self.onMethodCall(call: call, result: result)
    }
  }

  /// Handles all the supported Flutter method calls for the controlled
  /// `RtpSenderProxy`.
  func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
    let argsMap = call.arguments as? [String: Any]
    switch call.method {
    case "replaceTrack":
      let trackId = argsMap!["trackId"] as? String
      var track: MediaStreamTrackProxy?
      if trackId != nil {
        track = MediaStreamTrackStore.tracks[trackId!]
      } else {
        track = nil
      }

      self.rtpSender.replaceTrack(t: track)
      result(nil)
    case "getParameters":
      let encodings = self.rtpSender.getParameters().encodings
        .map { enc -> [String: Any] in
          [
            "rid": enc.rid,
            "active": enc.isActive,
            "maxBitrate": enc.maxBitrateBps,
            "maxFramerate": enc.maxFramerate,
            "scaleResolutionDownBy": enc.scaleResolutionDownBy,
            "scalabilityMode": enc.scalabilityMode,
          ]
        }

      result(["encodings": encodings])
    case "setParameters":
      let params = self.rtpSender.getParameters()
      let encodings = argsMap!["encodings"] as? [[String: Any]]?

      for e in encodings!! {
        let rid = e["rid"] as! String
        let enc = params.encodings.first(where: { $0.rid == rid })
        if enc == nil {
          result(FlutterError(
            code: "SenderError",
            message: "Could not set parameters: failed to find encoding with rid = " +
              rid,
            details: nil
          ))
          return
        }

        enc!.isActive = e["active"] as! Bool
        if let maxBitrate = e["maxBitrate"] as? Int {
          enc!.maxBitrateBps = NSNumber(value: maxBitrate)
        }
        if let maxFramerate = e["maxFramerate"] as? Double {
          enc!.maxFramerate = NSNumber(value: maxFramerate)
        }
        if let scaleResolutionDownBy = e["scaleResolutionDownBy"] as? Double {
          enc!.scaleResolutionDownBy = NSNumber(value: scaleResolutionDownBy)
        }
        if let scalabilityMode = e["scalabilityMode"] as? String {
          enc!.scalabilityMode = scalabilityMode
        }
      }

      self.rtpSender.setParameters(params: params)

      result(nil)
    case "dispose":
      self.channel.setMethodCallHandler(nil)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Converts this controller into a Flutter method call response.
  func asFlutterResult() -> [String: Any] {
    [
      "channelId": self.channelId,
    ]
  }
}

File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/controller/RtpSenderController.kt
package com.instrumentisto.medea_flutter_webrtc.controller

import com.instrumentisto.medea_flutter_webrtc.TrackRepository
import com.instrumentisto.medea_flutter_webrtc.proxy.RtpSenderProxy
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Controller of [RtpSenderProxy] functional.
 *
 * @param messenger Messenger used for creating new [MethodChannel]s.
 * @property sender Underlying [RtpSenderProxy] to perform [MethodCall]s on.
 */
class RtpSenderController(messenger: BinaryMessenger, private val sender: RtpSenderProxy) :
    MethodChannel.MethodCallHandler, IdentifiableController {
  /** Unique ID of the [MethodChannel] of this controller. */
  private val channelId = nextChannelId()

  /** Channel listened for the [MethodCall]s. */
  private val chan = MethodChannel(messenger, ChannelNameGenerator.name("RtpSender", channelId))

  init {
    chan.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "replaceTrack" -> {
        val trackId: String? = call.argument("trackId")
        val track =
            if (trackId != null) {
              TrackRepository.getTrack(trackId)!!
            } else {
              null
            }
        sender.replaceTrack(track)
        result.success(null)
      }
      "getParameters" -> {
        val encodings =
            sender.getParameters().encodings.map { enc ->
              mapOf(
                  "rid" to enc.rid,
                  "active" to enc.active,
                  "maxBitrate" to enc.maxBitrateBps,
                  "maxFramerate" to enc.maxFramerate,
                  "scaleResolutionDownBy" to enc.scaleResolutionDownBy,
                  "scalabilityMode" to enc.scalabilityMode,
              )
            }

        result.success(mapOf("encodings" to encodings))
      }
      "setParameters" -> {
        val params = sender.getParameters()
        val encodings: List<Map<String, Any>> = call.argument("encodings")!!

        for (e in encodings) {
          val rid = e["rid"] as String
          val enc = params.encodings.find { encoding -> encoding.rid == rid }
          if (enc == null) {
            result.error(
                "SenderException",
                "Could not set parameters: failed to find encoding with rid = $rid",
                null)
            return
          }

          enc.active = e["active"] as Boolean
          enc.maxBitrateBps = e["maxBitrate"] as Int?
          enc.maxFramerate = e["maxFramerate"] as Int?
          enc.scaleResolutionDownBy = e["scaleResolutionDownBy"] as Double?
          enc.scalabilityMode = e["scalabilityMode"] as String?
        }

        if (!sender.setParameters(params)) {
          result.error("SenderException", "Could not set parameters", null)
          return
        }

        result.success(null)
      }
      "dispose" -> {
        chan.setMethodCallHandler(null)
        result.success(null)
      }
    }
  }

  /**
   * Converts this [RtpSenderController] to the Flutter's method call result.
   *
   * @return [Map] generated from this controller which can be returned to the Flutter side.
   */
  fun asFlutterResult(): Map<String, Any> = mapOf("channelId" to channelId)
}
